providers/component-protocol/definition/cptype/type.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cptype
+
+// InitializeOperation .
+const (
+	// 协议定义的操作
+	// 用户通过URL初次访问
+	InitializeOperation OperationKey = "__Initialize__"
+	// 用于替换掉DefaultOperation，前端触发某组件，在协议Rending中定义了关联渲染组件，传递的事件是RendingOperation
+	RenderingOperation OperationKey = "__Rendering__"
+)
+
+// ComponentProtocol is protocol definition.
+type ComponentProtocol struct {
+	Version     string                   `json:"version" yaml:"version"`
+	Scenario    string                   `json:"scenario" yaml:"scenario"`
+	GlobalState *GlobalStateData         `json:"state" yaml:"state"`
+	Hierarchy   Hierarchy                `json:"hierarchy" yaml:"hierarchy"`
+	Components  map[string]*Component    `json:"components" yaml:"components"`
+	Rendering   map[string][]RendingItem `json:"rendering" yaml:"rendering"`
+}
+
+// GlobalStateData .
+type GlobalStateData map[string]interface{}
+
+// Hierarchy represents components' hierarchy.
+type Hierarchy struct {
+	Version string `json:"version" yaml:"version"`
+	Root    string `json:"root" yaml:"root"`
+	// structure的结构可能是list、map
+	Structure map[string]interface{} `json:"structure" yaml:"structure"`
+}
+
+// Component defines a component.
+type Component struct {
+	Version string `json:"version,omitempty" yaml:"version,omitempty"`
+	// 组件类型
+	Type string `json:"type,omitempty" yaml:"type,omitempty"`
+	// 组件名字
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+	// table 动态字段
+	Props interface{} `json:"props,omitempty" yaml:"props,omitempty"`
+	// 组件业务数据
+	Data ComponentData `json:"data,omitempty" yaml:"data,omitempty"`
+	// 前端组件状态
+	State map[string]interface{} `json:"state,omitempty" yaml:"state,omitempty"`
+	// 组件相关操作（前端定义）
+	Operations ComponentOps `json:"operations,omitempty" yaml:"operations,omitempty"`
+}
```

Comment:
aren't these struct defined in proto-go?

Reply:
Compatible for components already exist.
And it's better to have a layer to shield to handle type transfer between protobuf's `Value` with go's `interface{}`.